### PR TITLE
Cache plugin modules

### LINKED ISSUE #9029: [CT-3336] [Test Performance] Optimize the set_up_plugin_manager() function
### Short description

Profiling shows that the set_up_plugin_manager() function takes up 30% of the runtime on the main thread during integration testing. The hot path is set_up_plugin_manager() -> from_modules() -> iter_modules(). By caching the results we need after the first time iter_modules() is called, nearly all this time can be won back.

### Acceptance criteria

- The iter_modules() function is only called once in the above call tree for each test process.
- We have documented a notable improvement in performance in GitHub or via local profile.

### Impact to Other Teams

None. They may enjoy the speedier tests, though.

### Will backports be required?

No.

### PR DESCRIPTION
**Pull request by peterallenwebb**
resolves #9029

Local results:
- Before: make integration  1377.96s user 124.94s system 834% cpu 3:00.16 total
- After: make integration  1057.26s user 61.76s system 719% cpu 2:35.55 total

### Problem

Scanning all available python modules is slow.

### Solution

For the purposes of detecting dbt plugins, only do this once per process invocation and cache the result.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

**Review by ChenyuLInx — APPROVED** [review]
This is great!!!

**Comment by peterallenwebb** [comment]
Python 3.8 lacks `@functools.cache`, so we'll have to work around.

File: core/dbt/plugins/manager.py [hunk 1]
```diff
@@ -63,6 +65,17 @@ def get_manifest_artifacts(self, manifest: Manifest) -> PluginArtifacts:
         raise NotImplementedError(f"get_manifest_artifacts hook not implemented for {self.name}")
 
 
+@functools.cache
```
Comment (jtcohen6) [inline core/dbt/plugins/manager.py:68]: @peterallenwebb Per [docs](https://docs.python.org/3/library/functools.html#functools.cache), sounds like `@functools.cache` might just be 3.9+ syntactic sugar for `@functools.lru_cache(maxsize=None)`, which is available on earlier versions?

**Review by jtcohen6 — COMMENTED** [review]
very neat!!

**Pull request merged**